### ISSUE TITLE
Request for a demo

### ISSUE THREAD

**Issue by yanconglin** [kylevedder/zeroflow#4]
Hi, Kyleverdder,

Nice work! thank you for releasing the code.
Can you please provide a simple demo that takes as input two point clouds and outputs the flow?

Thank you.

**Comment by kylevedder** [comment]
Here is [argoverse2_tiny.zip](https://github.com/kylevedder/zeroflow/files/13059582/argoverse2_tiny.zip), a demo dataset with a single frame pair from the Argoverse2 `val` spit. It needs to be unzipped so that it appears as `/efs/argoverse2_tiny/` in the docker image, similar to the `/efs/argoverse2` setup described in the Getting Started documentation.

`argoverse2_tiny` contains three subfolders:

 - `argoverse2_tiny/val`: a single sequence with the single frame pair
 - `argoverse2_tiny/val_sceneflow`: the supervised ground truth for this frame pair
 - `argoverse2_tiny/val_supervised_out`: the output of the forward pass of our FastFlow3D 1x model on the `configs/fastflow3d/argo/supervised_tiny.py` config with our [supervised checkpoint](https://github.com/kylevedder/zeroflow_weights/blob/master/argo/supervised/supervised.ckpt), i.e.
 
 ```
 python test_pl.py configs/fastflow3d/argo/supervised_tiny.py --checkpoint /efs/sceneflow_golden_weights/argo/supervised/supervised.ckpt
 ```

Visualizing these results with 

```
python visualization/visualize_unsupervised_flow.py 
```

produces the following visualization of the ego vehicle (the center is depicted as the 1m red ball) driving across a bridge with a car driving in the opposite direction.

![2023-10-20-221519_1065x966_scrot](https://github.com/kylevedder/zeroflow/assets/6148088/725da9a0-81c1-472f-a58e-c408ca8c8b26)

Hope this provides everything you're looking for!

**Comment by yanconglin** [comment]
thank you and appreciate it!

**Comment by yanconglin** [comment]
Hi, again,
Could you please provide a minimal dataloader that loads the data (input and gt) from disk? What I am looking for is a minimal script to load the data (paired frames and gt) such that I can easily test my model on this dataset. I have followed [argoverse2-sf](https://github.com/kylevedder/argoverse2-sf) to process the downloader argoverse2 dataset.
Unfortunately, I am having a hard time reading the dataloaders in theis repo. 
Thank you in advance.

**Comment by kylevedder** [comment]
To be honest: no. The code is pretty readable, especially by research standards.

As an overview of the dataloader architecture:

The `argoverse_*_flow.py` and `waymo_*_flow.py` files provide pretty straightforward loading of dataset sequences -- `supervised` means using human annotations, `unsupervised` means using pseudolabels. A `*SequenceLoader` object loads `*Sequence` objects that wrap each Waymo Open or Argoverse 2 sequence. The `sequence_dataset.py` uses this standardized interface to turn the series of sequences into a single list of PyTorch'd matrix dictionaries for consumption by the neural networks.

For AV2, the supervised and unsupervised loaders both inherit from the `argoverse_raw_lidar.py` sequences / loader, which provides raw observations; the child objects just insert their own supervision keys.

**Comment by yanconglin** [comment]
Thank you for the pointer.  Always painful to handle those monster datasets with x TB data and complicated data structures. I managed to understand the dataloader somehow. 

Still, I would also like to ask: 
1. The dataloader does seem to be a per-file design (load an npz, do inference and then test, until all npz files are processed).  Out of curiosity, could you elaborate on this?
2. Wonder if you could kindly explain the ground removal on Waymo. There is no ground point label available AFAIK. I read the data preprocessing scripts but found little clue on the exact idea.


Appreciate it.

**Comment by kylevedder** [comment]
1. I don't understand what you are trying to ask. It loads two point clouds, and the `.npz` describing the flow between them.
2. I specify the Waymo Open version as 1.4.2 because it adds map information. Per the website:

>v1.4.2, March 2023: Added 3D map data as polylines or polygons. For 2D panoptic segmentation labels, added a mask to indicate pixels covered by multiple cameras.

**Comment by yanconglin** [comment]
Thank you for the reply. 
to clarify a bit:  I am surprised that the given dataloader is not reading all info (point clouds and flows) directly from the preprocessed npz files. It seems the pcds are still loaded from the original data, while the flow supervision is loaded from the processed data. Could you please shed some light on this? why not load the point clouds directly from the preprocessed data.

**Comment by kylevedder** [comment]
The AV2 `.npz` files are flow information. We do not duplicate the point cloud information because it's not necessary to do so and avoiding duplication saves space on disk.

**Comment by yanconglin** [comment]
Hi, again. 
Some details on Argo ground removal: how is this done?  I checked the chodosh et. al. paper and did not find much info. The only info I can find is from the challenge website:  "Contestants will be given as input a pair of sweeps, the vehicle ego-motion, and ground height information. "  It is not yet clear to me how exactly is this done. Could you please clarify? 

Good luck with ICLR rebuttal !

**Comment by kylevedder** [comment]
The AV2 Sensor dataset provides a rasterized heightmap.

The rasterized heightmap details the height of the ground in the global frame on a per-pillar basis. Point clouds are placed in the global frame, and each point's Z value is compared against the height of this heightmap; if the point's Z value is less than [40cm above the heightmap value](https://github.com/kylevedder/zeroflow/blob/c83f28737700b7cb87d8a1c48c138e2ef98e6b8b/dataloaders/argoverse_raw_lidar.py#L9), the point is marked as ground.

See [this function](https://github.com/kylevedder/zeroflow/blob/c83f28737700b7cb87d8a1c48c138e2ef98e6b8b/dataloaders/argoverse_raw_lidar.py#L126) for details.

For the AV2 LiDAR dataset, [we generate this rasterized heightmap from the vectorized map polygon geometry](https://github.com/kylevedder/zeroflow/blob/c83f28737700b7cb87d8a1c48c138e2ef98e6b8b/data_prep_scripts/argo/rasterize_heightmap.py) using kNN on the vector geometry vertices to get the height.
